src/fty_common_mlm_tntmlm.cc

```diff
@@ -49,129 +50,182 @@ MlmClient::~MlmClient()
-        connect();
+    if (!connected() && !connect()) {
+        return NULL;
+    }
+
+    if (!_poller) {
+        log_debug("poller not initialized");
+        return NULL;
     }
-    uint64_t wait = timeout;
-    if (wait > 300) {
-        wait = 300;
+
+    if (timeout_s > 300) { // 300 s. max
+        timeout_s = 300;
     }
 
-    uint64_t start          = static_cast<uint64_t>(zclock_mono());
-    uint64_t poller_timeout = wait * 1000;
+    int64_t quit_ms = zclock_mono() + (timeout_s * 1000);
 
     while (true) {
-        void* which = zpoller_wait(_poller, int(poller_timeout));
-        if (which == nullptr) {
-            log_warning("zpoller_wait (timeout = '%" PRIu64
-                        "') returned NULL. zpoller_expired == '%s', zpoller_terminated == '%s'",
-                poller_timeout, zpoller_expired(_poller) ? "true" : "false",
-                zpoller_terminated(_poller) ? "true" : "false");
-            return nullptr;
+        int timeleft_ms = int(quit_ms - zclock_mono());
+        if (timeleft_ms <= 0) {
+            log_debug("timeleft_ms: %dms", timeleft_ms);
+            break;
         }
-        uint64_t timestamp = static_cast<uint64_t>(zclock_mono());
-        if (timestamp - start >= poller_timeout) {
-            poller_timeout = 0;
-        } else {
-            poller_timeout = poller_timeout - (timestamp - start);
+
+        void* which = zpoller_wait(_poller, timeleft_ms);
+        if (!which) {
+            log_debug("zpoller_wait() returned NULL. (timeleft_ms: %dms, expired: %s, terminated: %s)",
+                timeleft_ms,
+                (zpoller_expired(_poller) ? "true" : "false"),
+                (zpoller_terminated(_poller) ? "true" : "false"));
+            break;
         }
+
         zmsg_t* msg = mlm_client_recv(_client);
-        if (!msg)
-            continue;
-        char* uuid_recv = zmsg_popstr(msg);
-        if (uuid_recv && uuid.compare(uuid_recv) == 0) {
+        if (msg) {
+            char* uuid_recv = zmsg_popstr(msg);
+            if (!(uuid_recv && streq(uuid_recv, uuid.c_str()))) {
+                zmsg_destroy(&msg);
+            }
             zstr_free(&uuid_recv);
-            return msg;
+            if (msg) {
+                return msg; // msg received
+            }
         }
-        zstr_free(&uuid_recv);
-        zmsg_destroy(&msg);
     }
+
+    return NULL; // timedout, terminated, ...
 }
 
-int MlmClient::sendto(const std::string& address, const std::string& subject, uint32_t timeout, zmsg_t** content_p)
+int MlmClient::sendto(const std::string& address, const std::string& subject, uint32_t timeout_s, zmsg_t** content_p)
 {
-    if (!connected()) {
-        connect();
+    if (!connected() && !connect()) {
+        return -1;
     }
-    return mlm_client_sendto(_client, address.c_str(), subject.c_str(), nullptr, timeout, content_p);
+
+    int r = mlm_client_sendto(_client, address.c_str(), subject.c_str(), NULL, timeout_s * 1000, content_p);
```

Comment:
No need to create temporally variable

Reply:
Yes, it's not needed; but this is for readabilty...